pricegraph/src/graph/shortest_paths.rs

```diff
@@ -0,0 +1,311 @@
+    pub fn connected_nodes(&self) -> Vec<G::NodeId> {
+        let mut node_indices: Vec<_> = self
+            .predecessor_store
+            .predecessors
+            .iter()
+            .enumerate()
+            .filter_map(|(i, &pre)| pre.map(|_| self.graph.from_index(i)))
+            .collect();
+        debug_assert!(!node_indices.contains(&self.source));
+        node_indices.push(self.source);
+        node_indices
+    }
+}
+
+impl<G> ShortestPathGraph<G>
+where
+    G: IntoNodeIdentifiers + IntoEdges + NodeIndexable,
+    G::EdgeWeight: FloatMeasure,
+{
+    /// Initializes a shortest path graph that will be later built with the
+    /// Bellman-Ford algorithm.
+    fn empty(g: G, source: G::NodeId) -> Self {
+        let predecessors = vec![None; g.node_bound()];
+        let mut distance = vec![<_>::infinite(); g.node_bound()];
+        distance[g.to_index(source)] = <_>::zero();
+
+        let predecessor_store = PredecessorStore {
+            predecessors,
+            distance,
+        };
+
+        ShortestPathGraph {
+            graph: g,
+            predecessor_store,
+            source,
+        }
+    }
+
+    /// Checks for negative weight cycle and, if any is found, creates loop in
+    /// the predecessor store
+    fn mark_cycle(&mut self) -> Option<G::NodeId> {
+        // This is the last step of the Bellman-Ford algorithm. It tries to relax
+        // each node: if a node can be relaxed then a negative cycle exists and is
+        // created in the predecessor store; otherwise no such cycle exists.
+
+        for i in self.graph.node_identifiers() {
+            for edge in self.graph.edges(i) {
+                let j = edge.target();
+                let w = *edge.weight();
+
+                if *self.distance(i) + w < *self.distance(j) {
+                    self.update_predecessor(j, Some(i));
+                    return Some(j);
+                }
+            }
+        }
+        None
+    }
+
+    /// Returns a negative cycle, if it exists.
+    fn find_cycle(&mut self) -> Option<NegativeCycle<G::NodeId>> {
+        let search_start = match self.mark_cycle() {
+            Some(node) => node,
+            None => return None,
+        };
+
+        find_cycle(
+            self.graph,
+            &self.predecessor_store.predecessors,
+            search_start,
+            None,
+        )
+    }
+}
+
+impl<G> ShortestPathGraph<G>
+where
+    G: NodeCount + IntoNodeIdentifiers + IntoEdges + NodeIndexable,
+    G::EdgeWeight: FloatMeasure,
+{
+    /// Creates a representation of all shortest paths from the given source
+    /// to any other node in the graph.
+    ///
+    /// Shortest paths are well defined if and only if the graph does not
+    /// contain any negative weight cycle reachabe from the source. If a
+    /// negative weight cycle is detected, it is returned as an error.
+    pub fn new(g: G, source: G::NodeId) -> Result<Self, NegativeCycle<G::NodeId>> {
+        bellman_ford(g, source)
```

Comment:
Very nice! We can now completely abstract away the actual search algorithm and instead just provide an interface to "finding the shortest path in a graph" :+1: